drivers/clock_control/nrf_power_clock.c

```diff
@@ -12,269 +12,308 @@
 #include <drivers/clock_control.h>
 #include <sys/__assert.h>
 #include <nrf_clock.h>
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-#include <nrf_power.h>
 #include <drivers/clock_control/nrf_clock_control.h>
-#endif
+#include <nrf_power.h>
+#include <logging/log.h>
+#include <sensor.h>
+
+LOG_MODULE_REGISTER(clock_control, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+/* Helper logging macros which prepends device name to the log. */
+#define CLOCK_LOG(lvl, dev, ...) \
+	LOG_##lvl("%s: " GET_ARG1(__VA_ARGS__), dev->config->name \
+			COND_CODE_0(NUM_VA_ARGS_LESS_1(__VA_ARGS__),\
+					(), (, GET_ARGS_LESS_1(__VA_ARGS__))))
+#define ERR(dev, ...) CLOCK_LOG(ERR, dev, __VA_ARGS__)
+#define WRN(dev, ...) CLOCK_LOG(WRN, dev, __VA_ARGS__)
+#define INF(dev, ...) CLOCK_LOG(INF, dev, __VA_ARGS__)
+#define DBG(dev, ...) CLOCK_LOG(DBG, dev, __VA_ARGS__)
+
+/* returns true if clock stopping must be deferred due to active calibration. */
+typedef void (*nrf_clock_stop_handler_t)(struct device *dev);
+
+/* Clock instance structure */
+struct nrf_clock_control {
+	atomic_t ref;		/* Users counter */
+	sys_slist_t list;	/* List of users requesting callback */
+	bool started;		/* When set clock is started */
+};
+
+/* Clock instance static configuration */
+struct nrf_clock_control_config {
+	u32_t start_tsk;	/* Clock start task */
+	u32_t started_evt;	/* Clock started event */
+	u32_t stop_tsk;		/* Clock stop task */
+	nrf_clock_stop_handler_t stop_handler; /* Function called before stop */
+};
 
-static u8_t m16src_ref;
-static u8_t m16src_grd;
-static u8_t k32src_initialized;
+#define INT_MASK \
+	(NRF_CLOCK_INT_HF_STARTED_MASK | NRF_CLOCK_INT_LF_STARTED_MASK \
+	COND_CODE_1(CONFIG_USB_NRF52840, \
+		    (|NRF_POWER_INT_USBDETECTED_MASK | \
+		      NRF_POWER_INT_USBREMOVED_MASK | \
+		      NRF_POWER_INT_USBPWRRDY_MASK), ()))
 
-static int m16src_start(struct device *dev, clock_control_subsys_t sub_system)
+/* Function returns current interrupt priority or 0xff if in thread mode */
+static int z_arch_irq_current_prio(void)
```

Comment:
This function would seem quite out of place here. Shouldnt it be in arch/arm/core/irq_mange.c  or so?

Reply:
yes, that was my forgotten intention. See #17301.